admin/documentation-updates (#20)
CouncilDataProject/cookiecutter-cdp-deployment: README.md

| hosting_github_repo_name       | A specific name to give to the repository. (Used in the web application's full address)                                    | cdp-seattle                           | king-county                              |
| hosting_github_url             | From the provided information, the expected URL of the web application.                                                    | jacksonmaxfield.github.io/cdp-seattle | councildataproject.github.io/king-county |

### Things to Know

Much of CouncilDataProject processing and resource management can be handled for
free and purely on GitHub. However we do rely on a select few resources outside
of GitHub to manage all services and applications.

The only service that will require a billing account to manage payment for
resources used, is [Google Cloud](#google-cloud).
Google Cloud will manage all databases, file storage, and (if needed)
[speech-to-text](#speech-to-text) for transcription. You can see more about the
average monthly cost of running a CDP Instance in [Cost](#cost).

> [Pulumi](#pulumi) is a service to manage and track infrastructure deployment state.

Reviewer: It looks like there isn't a Pulumi heading in this README, so this link doesn't go to anything

-   Continuous integration
    -   Preconfigured for your web application to fully deploy
    -   Preconfigured to deploy all required CDP infrastructure
    -   Preconfigured to run CDP pipelines using GitHub Actions

To generate a new repository from this template,
in a terminal with Python 3.5+ installed, run:

```bash
pip install cookiecutter
cookiecutter gh:CouncilDataProject/cookiecutter-cdp-deployment
```

_Note: This will only create the basic repository. You will still need to setup_
_Google Cloud and Pulumi accounts._

### Google Cloud

All of your deployments data and some data processing will be done using
Google Cloud Platform (GCP).

-   Your deployment's provided and generated data (meeting dates,
    committee names, councilmember details, etc) will live in
    [Firestore](https://cloud.google.com/firestore).
-   Your deployment's generated files (audio clips, transcripts, etc.)
    will live in [Filestore](https://cloud.google.com/filestore).
-   When provided a video without closed captions, the audio from the provided video
    will be processed using [Speech-to-Text](https://cloud.google.com/speech-to-text).

> All of these resources will be set up for you using [Pulumi](#pulumi) but you will

Reviewer: Same as above, this Pulumi link doesn't go anywhere

need to create both Google Cloud and Pulumi accounts. More information on these
services and the steps for account creation can be found in the generated
repository's README.

## Cost

CDP was created and maintained by a group of people working on it in their free time.
We didn't want to pay extreme amounts of money so why should you?

To that end, we try to make CDP as low cost as possible.
Many of the current features are entirely free as long as the repo is open source:

Free Resources and Infrastructure:

-   Event Processing (GitHub Actions)
-   Event and Legislation Indexing (GitHub Actions)
-   Web Hosting (GitHub Pages)
-   Infrastructure State Management (Pulumi)

The backend resources and processing are the only real costs and depend on usage.
The more users that use your web application, the more the database and file storage

> cost. The CDP-Seattle monthly averages below are the averages for the most utilized
> months of it's existance so take these as close to upper-bounds.

Reviewer: Minor grammar suggestion:

```suggestion
cost. The CDP-Seattle monthly averages below are for the most utilized
months of its existence so take these as close to upper-bounds.
```

Billed Resources and Infrastructure:

-   [Cloud Firestore Pricing](https://firebase.google.com/pricing/)
    _CDP-Seattle monthly average: ~$8.00_
-   [Google Storage Pricing](https://cloud.google.com/storage/pricing#price-tables)
    _CDP-Seattle monthly average: ~$3.00_
-   [Google Speech-to-Text Pricing](https://cloud.google.com/speech-to-text/pricing)
    _CDP-Seattle monthly average: ~$22.00_

**Total Average Monthly Cost**: $33.00

### Speech-to-Text

You may not need to use speech-to-text! In the case your municipalicity provides closed
caption files in a format we support parsing and cleaning, we can use those files
instead of using speech-to-text. When using closed caption files for transcription

> generation. CDP-Seattle speech-to-text costs dropped to ~$2.00 / month.

Reviewer: ```suggestion
generation, CDP-Seattle speech-to-text costs dropped to ~$2.00 / month.
```

To use closed captions files instead of generating a transcript from audio,
your event gather function can attach a
[`closed_caption_uri`](https://councildataproject.github.io/cdp-backend/cdp_backend.pipeline.html#cdp_backend.pipeline.ingestion_models.Session)
to the `Session` object.

With speech-to-text cost removed, total average monthly cost for CDP-Seattle is ~$13.00.

### Future Processing Features

As we add more features to CDP that require additional processing or resources we will
continue to try to minimize their costs wherever possible. Further, if a feature is
optional, we will create a flag that maintainers can set to include or exclude the
additional processing or resource usage.
See [Upgrades and New Features](#upgrades-and-new-features) for more information.

## Upgrades and New Features

In general, all updates and upgrades are handled easily for you through
automated systems that run using GitHub Actions on your repository.

Upgrades are delivered either just before processing time (during package installation)
or as a part of weekly system checks and auto-deployments.

#### Backend Pipeline Upgrades

> Due to how we have constructed how event processing works, every time CDP
> pipelines are initiated, they also install the latest non-breaking-changes version of
> the [cdp-backend](https://github.com/CouncilDataProject/cdp-backend) package.

Reviewer: ```suggestion
Every time CDP pipelines are initiated, event processing will also install the latest non-breaking-changes version of the [cdp-backend](https://github.com/CouncilDataProject/cdp-backend) package.
```

In this way, every time your pipeline runs to gather events, index events, or
index pieces of legislation, you can be sure you are up-to-date.

#### Backend Infrastructure Upgrades

Every week, your repository will automatically check the status of your infrastructure
and deploy updates as detected.

#### Frontend Web Application Upgrades

Every week, your repository will automatically build and deploy the CDP web application,
during the build process any non-breaking changes and new features will automatically be
pulled in.

### Breaking Changes

In the case we release a new version of
[cdp-backend](https://github.com/CouncilDataProject/cdp-backend) or
[cdp-frontend](https://github.com/CouncilDataProject/cdp-frontend) that includes
breaking changes we will also release an upgrade guide that you will be able to find
in this repository.

> In the case we can fully automate the upgrade, we will include a script to do so that
> CDP Instance maintainers will simply have to run.

Reviewer: Minor edit for readability, but I'm not sure where a maintainer might run an upgrade script, so feel free to edit that part. 

```suggestion
In the case we can fully automate the upgrade, we will include a script that
CDP Instance maintainers will simply have to run on their project.
```

months of it's existance so take these as close to upper-bounds.

-   [Cloud Firestore Pricing](https://firebase.google.com/pricing/)
    _CDP-Seattle monthly average: ~$8.00_
-   [Google Storage Pricing](https://cloud.google.com/storage/pricing#price-tables)
    _CDP-Seattle monthly average: ~$3.00_
-   [Google Speech-to-Text Pricing](https://cloud.google.com/speech-to-text/pricing)
    _CDP-Seattle monthly average: ~$22.00_

**Total Average Monthly Cost**: $33.00

### Speech-to-Text

You may not need to use speech-to-text! In the case your municipalicity provides closed
caption files in a format we support parsing and cleaning, we can use those files
instead of using speech-to-text. When using closed caption files for transcription
generation. CDP-Seattle speech-to-text costs dropped to ~$2.00 / month because an
occasional meeting didn't have closed captions. You can attach a
[`closed_caption_uri`](https://councildataproject.github.io/cdp-backend/cdp_backend.pipeline.html#cdp_backend.pipeline.ingestion_models.Session)
to the `Session` object during event ingestion.

With speech-to-text cost removed, total average monthly cost for CDP-Seattle is ~$11.00.

> ### Future Processing Features

Reviewer: It might be worth noting how updates happen - do they happen in `cdp-backend` and therefore are automatically applied to my deployment? Do I get notified somehow?

Author: The reason for the having the issue template is that I can see it feasible to automate the PR to add the deployment details.

if you have a template with the repo name, list of maintainers, supported municipality, and some other metadata, you can create a bot to automatically add those details to the list and create a PR.

(Overkill probably but it would be neat)

Author: Also we do have a slightly built up system for pulling in upgrades already. In the generated repos, Makefile there is a command called `update-from-cookiecutter` that reruns the cookiecutter in place. It works pretty well but you have to be good about git because after it runs you as a maintainer need to cherrypick which changes to keep and to discard.

It has worked so far in maintaining the CouncilDataProject/example repo.

index pieces of legislation, you can be sure you are up-to-date.

#### Backend Infrastructure Upgrades

Every week, your repository will automatically check the status of your infrastructure
and deploy updates if needed.

#### Frontend Web Application Upgrades

Every week, your repository will automatically build and deploy the CDP web application,
during the build process any non-breaking changes and new features will automatically be
pulled in.

### Breaking Changes

In the case we release a new version of
[cdp-backend](https://github.com/CouncilDataProject/cdp-backend) or
[cdp-frontend](https://github.com/CouncilDataProject/cdp-frontend) that includes
breaking changes we will also release an upgrade guide that you will be able to find
in this repository.

In the case we can fully automate the upgrade, we will include a script to do so that
CDP instance maintainers will simply have to run.

### Notification of Updates and Release Notes

We are still in the process of developing a notification system for alerting CDP
instance maintainers that a new version of `cdp-backend` or `cdp-frontend` was
released. These notifications will include details on new features, bugfixes, and

> if there is anything the instance maintainer needs to do to enable the release.

Reviewer: Maybe add a link to become a watcher of the two repos so you can be notified of releases?

Author: This is a much simpler system and I appreciate this. :joy:

Reviewer: 🤔 I'm not sure if there is a direct link to watch something on GitHub actually

Author: I don't think there is either, but I think we can include instructions that say "Click watch, select custom, and select Release, then Apply." You will receive emails when that repository releases new versions,
